_posts/2019-04-06-omakase-relay.markdown

```diff
@@ -0,0 +1,245 @@
+Relay terms, this is called [data masking][masking] and it is one of the first hurdles for someone new to Relay to
+[grok][]. In REST clients, and GraphQL API clients like Apollo Client, you make a request and that request is
+passed through the React tree. E.g.
+
+{% include epic_img.html url="/images/omakase-relay/tree.png" title="REST inspired props" style="width:100%;" %}
+
+This means most components know more about the request than it probably needs, as it may be needed to pass on to
+the component's children. This can lead to over-fetching, or even worse, not knowing if you can delete or refactor
+a component.
+
+Data masking solves this by hiding data that the component didn't request. I've still yet to find the right visual
+abstraction, but I feel this just about pays for itself.
+
+{% include epic_img.html url="/images/omakase-relay/isolation.png" title="Relay isolation tree" style="width:100%;" %}
+
+You let Relay be responsible for consolidating all your fragments into a query, request it, and (mostly) passing it
+through your component hierarchy. This means Relay powered components can be safely changed and drastically reduces
+the chance for unintended consequences elsewhere.
+
+This isolation gives Artsy engineers the safety to work on projects with tens of contributors, which change over
+long time periods without accruing technical debt. The components we create are nearly all focused only on the
+data-driven aspects of rendering a GraphQL response into views.
+
+## Co-location
+
+Relay helped us move to one file representing everything a component needed. Effectively a single file now handles
+the styles, the actual view content hierarchy, and the exact parts of the API it needs to render itself.
+
+<img src="/images/omakase-relay/co-location.png">
+
+In roughly that proportion too, though our most modern code uses the Artsy design system [Palette][palette] which
+drastically reduces the need for style in our components.
+
+Co-location's biggest selling point is reducing [cognitive load](https://en.wikipedia.org/wiki/Cognitive_load),
+having everything you need in one place makes it easier to understand how a component works. This makes code review
+simpler, and lowers the barrier to understanding the entire systems at scale.
+
+## Community
+
+When we adopted Relay, there was no competition - we'd have just used the `fetch` API. Over time, [the Apollo
+team][apollo] came up and really put a considerable amount of effort into lowering the barriers to entry, and
+making it feasible to build complex apps easily.
+
+As we hired a set of new engineers, the "Apollo vs Relay" debate came up. Interested in whether we would still
+start with Relay today, we ran an audit last year of what it would take to re-create a lot of the infrastructure we
+love in Relay atop of the (much more popular) Apollo GraphQL eco-system and saw it was feasible but would require a
```

Comment:
At its core, Apollo remains a lib that does more at runtime vs Relay. I don’t think we need to go into that detail, but I also wouldn’t say we could get 100% the same thing if we built enough atop of Apollo.

```suggestion
love in Relay atop of the (much more popular) Apollo GraphQL eco-system and saw it was reasonably feasible but would require a
```